src/plugins/kibana_usage_collection/server/collectors/cloud/detector/cloud_service.ts

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import fs from 'fs';
+import { isObject, isString } from 'lodash';
+import defaultRequest from 'request';
+import type { OptionsWithUri, Response as DefaultResponse } from 'request';
+import { CloudServiceResponse } from './cloud_response';
+
+/** @internal */
+export type Request = typeof defaultRequest;
+
+/** @internal */
+export type RequestOptions = OptionsWithUri;
+
+/** @internal */
+export type Response = DefaultResponse;
+
+/** @internal */
+export interface CloudServiceOptions {
+  _request?: Request;
+  _fs?: typeof fs;
+  _isWindows?: boolean;
+}
+
+/**
+ * CloudService provides a mechanism for cloud services to be checked for
+ * metadata that may help to determine the best defaults and priorities.
+ */
+export class CloudService {
+  private readonly _request: Request;
+  protected readonly _name: string;
+
+  constructor(name: string, options: CloudServiceOptions = {}) {
+    this._name = name.toLowerCase();
+
+    // Allow the HTTP handler to be swapped out for tests
+    const { _request = defaultRequest } = options;
+
+    this._request = _request;
+  }
+
+  /**
+   * Get the search-friendly name of the Cloud Service.
+   */
+  getName() {
+    return this._name;
+  }
+
+  /**
+   * Using whatever mechanism is required by the current Cloud Service,
+   * determine if Kibana is running in it and return relevant metadata.
+   */
+  checkIfService() {
+    return this._checkIfService(this._request).catch(() => this._createUnconfirmedResponse());
```

Comment:
NIT: I wonder if we should handle _known_ retryable errors 🤔 

Since we're calling it only once, during start-up. Unless the APIs we use have a high-enough SLA that we don't really need to worry about those errors at all.

What do you think?

Reply:
> Unless the APIs we use have a high-enough SLA that we don't really need to worry about those errors

I couldn't find any SLA info on the metadata APIs[0], however it should be noted that these metadata IPs are all [link-local addresses](https://en.wikipedia.org/wiki/Link-local_address): they're only accessible from the instance itself, and the APIs are on the same physical host that's running the VM, making network failures less likely (though not impossible).

Retries would certainly make this more bulletproof, however I'd prefer to treat it as an enhancement to follow up on later if you don't feel too strongly about it. WDYT?

[0] Docs for each of the APIs:
https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service
https://cloud.google.com/compute/docs/storing-retrieving-metadata
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html

Reply:
Makes sense! Thanks for the docs! We can go ahead and revisit later :)